### PR TITLE
Vimc 3548 - Humungous brutal cull of test csvs.

### PR DESCRIPTION
**Pull request by weshinsley**
A fairly cumbersome PR, which refactors tests, and is intended not to add any functional behaviour. (There are a couple of small changes to the main codebase, where the refactored tests failed, and the best solution was to fix the main code)

All CSVs for testing are now created on the fly in a temp directory, which makes it easier to see what the test is doing, rather than examining over 200 sample CSV files. It also allows a bit of factoring out common bits of behaviour, such as creating 1 or 2 "standard" touchstones, or the demography entries needed for testing touchstone_demographic_data.

Also, the db work is much faster now - I am caching the connection, and working "persistently" on the database, looking up serial ids rather than assuming they start at 1. This probably makes for some more robust tests anyway.

### CODE REVIEW

File: tests/testthat/helper-tests.R [hunk 1]
```diff
@@ -2,7 +2,40 @@ test_path <- function(context, path) {
   file.path("examples", context, path)
 }
 
+cache_con <<- NULL
```
Comment (richfitz) [inline tests/testthat/helper-tests.R]: I don't believe you want the double arrow here, just within the functions.  What I usually do though is:

```
cache <- new.env(parent = emptyenv())
```

Then use this as

```
if (is.null(cache$con)) {
  cache$con <- new_connection()
}
cache$con
```

etc

File: tests/testthat/helper-tests.R [hunk 2]
```diff
@@ -38,6 +71,147 @@ test_compare_csv_db <- function(con, csv, db) {
...
 }
 
+compare_csv <- function(res, tables) {
+  expect_true(all(unlist(lapply(seq_along(tables), function(ti) {
+    table <- tables[ti]
+    test_compare_csv_db(
+      res$con, res$e[[paste0(table, "_csv")]], table)}))))
+}
+
+mess_with <- function(path, csv, col, row, text) {
```
Comment (richfitz) [inline tests/testthat/helper-tests.R:81]: So, not the Wu Tang Clan then?

File: tests/testthat/helper-tests.R [hunk 3]
```diff
@@ -38,6 +71,147 @@ test_compare_csv_db <- function(con, csv, db) {
...
+  if (row == 0) {
+    names(data)[names(data) == col] <- text
+  } else {
+    data[[col]][row] <- text
+  }
+  write.csv(data, file.path(path, "meta", csv), row.names = FALSE)
+}
+
+db_file <- function(db, f) {
+  if (!db) f
```
Comment (richfitz) [inline tests/testthat/helper-tests.R]: As a style point I usually do this either with braces or on exactly one line as

```
if (!db) f else paste0("db_", f)
```

File: tests/testthat/test_touchstone_country.R [hunk 4]
```diff
@@ -12,82 +12,122 @@ test_touchstone_country <- function(test_name, ...) {
...
-               class = "simpleError")
-})
+# For the next two tests, I need to work out how to read a single
+# line (plus header) CSV that contains both NA and "" - and I want
+# to keep the "" as "", and the NA as numerical NA. R seems to want
+# to convert "" to NA, unless I read.csv with colClasses='character'
+# but then it's a pain to convert "NA" back to NA again at specific
+# points. See read_meta in utils.R
+
+#test_that("Empty disease in CSV", {
```
Comment (richfitz) [inline tests/testthat/test_touchstone_country.R]: Was this still wanted?

**Review by richfitz — CHANGES_REQUESTED** [review]
What a heroic effort. I see a couple of tiny things to look at and you probably want to fix the missed lines, especially if the aim was not to change functionality as there is now nontrivial package code being missed ( https://codecov.io/gh/vimc/stoner/src/vimc-3548/R/scenario_description.R#L69, https://codecov.io/gh/vimc/stoner/src/vimc-3548/R/touchstone_countries.R#L33, https://codecov.io/gh/vimc/stoner/src/vimc-3548/R/touchstone_demographic_dataset.R#L218)

File: tests/testthat/helper-tests.R (hunk 1 above)
Comment (weshinsley) [reply tests/testthat/helper-tests.R]: Thanks - have gone for the better way!

File: tests/testthat/helper-tests.R (hunk 3 above)
Comment (weshinsley) [reply tests/testthat/helper-tests.R]: Yep, that's nicer.

File: tests/testthat/test_touchstone_country.R (hunk 4 above)
Comment (weshinsley) [reply tests/testthat/test_touchstone_country.R]: Fixed in utils.R:read_meta. We want columns to be auto-detected - except disease and country which are always string (even "") - and focal_coverage_set, which is always numeric. Realistically, these two are slightly over-the-top tests...

**Review by richfitz — APPROVED** [review]

**Pull request merged**